diff --git a/lib/libriscv/memory.cpp b/lib/libriscv/memory.cpp
index 79f0cc1a..b2467976 100644
--- a/lib/libriscv/memory.cpp
+++ b/lib/libriscv/memory.cpp
@@ -3,7 +3,7 @@
 #include "decoder_cache.hpp"
 #include "internal_common.hpp"
 #include <inttypes.h>
-#if defined(__linux__) || defined(__FreeBSD__) || defined(__wasm__)
+#if defined(__linux__) || defined(__FreeBSD__) || defined(__wasm__) || defined(__APPLE__)
 #define DEMANGLE_ENABLED
 #include <sys/mman.h>
 #ifndef MAP_NORESERVE
@@ -35,7 +35,7 @@ namespace riscv
 
 			if (options.use_memory_arena)
 			{
-#if defined(__linux__) || defined(__FreeBSD__)
+#if defined(__linux__) || defined(__FreeBSD__) || defined(__APPLE__)
 				if constexpr (encompassing_Nbit_arena != 0)
 				{
 					static_assert(flat_readwrite_arena || encompassing_Nbit_arena == 0,
@@ -158,7 +158,7 @@ namespace riscv
 		this->evict_execute_segments();
 		// only the original machine owns arena
 		if (this->m_arena.data != nullptr && !is_forked()) {
-#if defined(__linux__) || defined(__FreeBSD__)
+#if defined(__linux__) || defined(__FreeBSD__) || defined(__APPLE__)
 			if constexpr (riscv::encompassing_Nbit_arena != 0)
 			{
 				// munmap() the entire address space
