diff --git a/blink/blink.c b/blink/blink.c
index 2c889d6..e2c2582 100644
--- a/blink/blink.c
+++ b/blink/blink.c
@@ -181,0 +182 @@ static void OnFatalSystemSignal(int sig, siginfo_t *si, void *ptr) {
+  u64 pte;
@@ -183 +184 @@ static void OnFatalSystemSignal(int sig, siginfo_t *si, void *ptr) {
-  sig = FixXnuSignal(m, sig, si);
+  sig = FixXnuSignal(m, sig, si, &pte);
@@ -190 +191 @@ static void OnFatalSystemSignal(int sig, siginfo_t *si, void *ptr) {
-  if (IsSelfModifyingCodeSegfault(m, si)) return;
+  if (IsSelfModifyingCodeSegfault(m, si, pte)) return;
diff --git a/blink/blinkenlights.c b/blink/blinkenlights.c
index fac1ff0..625a5d3 100644
--- a/blink/blinkenlights.c
+++ b/blink/blinkenlights.c
@@ -3654,0 +3655 @@ static void OnSigSegv(int sig, siginfo_t *si, void *uc) {
+    u64 pte;
@@ -3657 +3658 @@ static void OnSigSegv(int sig, siginfo_t *si, void *uc) {
-  sig = FixXnuSignal(m, sig, si);
+  sig = FixXnuSignal(m, sig, si, &pte);
@@ -3660 +3661 @@ static void OnSigSegv(int sig, siginfo_t *si, void *uc) {
-  if (IsSelfModifyingCodeSegfault(m, si)) return;
+  if (IsSelfModifyingCodeSegfault(m, si, pte)) return;
diff --git a/blink/flag.c b/blink/flag.c
index cdc88ce..1d01ba4 100644
--- a/blink/flag.c
+++ b/blink/flag.c
@@ -28,0 +29 @@ bool FLAG_alsologtostderr;
+bool FLAG_hacknofaults = true;
diff --git a/blink/flag.h b/blink/flag.h
index 2347b22..f51ca64 100644
--- a/blink/flag.h
+++ b/blink/flag.h
@@ -12,0 +13 @@ extern bool FLAG_alsologtostderr;
+extern bool FLAG_hacknofaults;
diff --git a/blink/machine.h b/blink/machine.h
index 5fd0d9f..1176455 100644
--- a/blink/machine.h
+++ b/blink/machine.h
@@ -552 +552 @@ void HandleFatalSystemSignal(struct Machine *, const siginfo_t *);
-bool IsSelfModifyingCodeSegfault(struct Machine *, const siginfo_t *);
+bool IsSelfModifyingCodeSegfault(struct Machine *, const siginfo_t *, u64 pte);
@@ -554 +554 @@ bool IsSelfModifyingCodeSegfault(struct Machine *, const siginfo_t *);
-int FixXnuSignal(struct Machine *, int, siginfo_t *);
+int FixXnuSignal(struct Machine *, int, siginfo_t *, u64 *pte);
diff --git a/blink/smc.c b/blink/smc.c
index ec04e87..baceaa1 100644
--- a/blink/smc.c
+++ b/blink/smc.c
@@ -18,0 +19 @@
+#include <stdio.h>
@@ -131,2 +132 @@ i64 ProtectRwxMemory(struct System *s, i64 rc, i64 virt, i64 size,
-bool IsSelfModifyingCodeSegfault(struct Machine *m, const siginfo_t *si) {
-  u64 pte;
+bool IsSelfModifyingCodeSegfault(struct Machine *m, const siginfo_t *si, u64 pte) {
@@ -137,0 +138,2 @@ bool IsSelfModifyingCodeSegfault(struct Machine *m, const siginfo_t *si) {
+  
+  #ifndef __APPLE__
@@ -138,0 +141,4 @@ bool IsSelfModifyingCodeSegfault(struct Machine *m, const siginfo_t *si) {
+  #else
+  vaddr = m->faultaddr;
+  #endif
+  
@@ -139,0 +146 @@ bool IsSelfModifyingCodeSegfault(struct Machine *m, const siginfo_t *si) {
+  if(!FLAG_hacknofaults)
diff --git a/blink/xnu.c b/blink/xnu.c
index 24d37c0..26bfc2b 100644
--- a/blink/xnu.c
+++ b/blink/xnu.c
@@ -22,2 +22 @@
-int FixXnuSignal(struct Machine *m, int sig, siginfo_t *si) {
-  u64 pte;
+int FixXnuSignal(struct Machine *m, int sig, siginfo_t *si, u64 *pte) {
@@ -31 +30,2 @@ int FixXnuSignal(struct Machine *m, int sig, siginfo_t *si) {
-    ConvertHostToGuestAddress(m->system, si->si_addr, &pte);
+    if (FLAG_hacknofaults) pte = 0;
+    m->faultaddr = ConvertHostToGuestAddress(m->system, si->si_addr, pte);
@@ -35,3 +35,3 @@ int FixXnuSignal(struct Machine *m, int sig, siginfo_t *si) {
-    if ((pte & PAGE_V) &&
-        ((pte & (PAGE_U | PAGE_RW)) != (PAGE_U | PAGE_RW) ||
-         (pte & (PAGE_U | PAGE_RW | PAGE_XD)) == (PAGE_U | PAGE_RW))) {
+    if (FLAG_hacknofaults||((*pte & PAGE_V) &&
+        ((*pte & (PAGE_U | PAGE_RW)) != (PAGE_U | PAGE_RW) ||
+         (*pte & (PAGE_U | PAGE_RW | PAGE_XD)) == (PAGE_U | PAGE_RW)))) {
